src/Mvc/GetDocumentInsider/src/Commands/GetDocumentCommandWorker.cs

```diff
@@ -22,200 +46,325 @@ public static int Process(GetDocumentCommandContext context)
-                    Reporter.WriteWarning(Resources.FormatMethodNotFound(methodName, serviceName));
-                    return false;
+                    break;
                 }
-                else if (!typeof(Task).IsAssignableFrom(method.ReturnType))
+            }
+
+            if (serviceType == null)
+            {
+                Reporter.WriteError(Resources.FormatServiceTypeNotFound(DocumentService));
+                return false;
+            }
+
+            var getDocumentsMethod = GetMethod(
+                GetDocumentsMethodName,
+                serviceType,
+                GetDocumentsParameterTypes,
+                GetDocumentsReturnType,
+                reportErrors: true,
+                isStatic: false);
+            if (getDocumentsMethod == null)
+            {
+                return false;
+            }
+
+            var generateMethod = GetMethod(
+                GenerateMethodName,
+                serviceType,
+                GenerateMethodParameterTypes,
+                GenerateMethodReturnType,
+                reportErrors: true,
+                isStatic: false);
+            if (generateMethod == null)
+            {
+                return false;
+            }
+
+            var service = services.GetService(serviceType);
+            if (service == null)
+            {
+                Reporter.WriteError(Resources.FormatServiceNotFound(DocumentService));
+                return false;
+            }
+
+            var documentNames = (IEnumerable<string>)InvokeMethod(getDocumentsMethod, service, GetDocumentsArguments);
+            if (documentNames == null)
+            {
+                return false;
+            }
+
+            // Write out the documents.
+            Directory.CreateDirectory(context.OutputDirectory);
+            var filePathList = new List<string>();
+            foreach (var documentName in documentNames)
+            {
+                var filePath = GetDocument(
+                    documentName,
+                    context.ProjectName,
+                    context.OutputDirectory,
+                    generateMethod,
+                    service);
+                if (filePath == null)
                 {
-                    Reporter.WriteWarning(Resources.FormatMethodReturnTypeUnsupported(
-                        methodName,
-                        serviceName,
-                        method.ReturnType,
-                        typeof(Task)));
                     return false;
                 }
 
-                var service = services.GetService(serviceType);
-                if (service == null)
+                filePathList.Add(filePath);
+            }
+
+            // Write out the cache file.
+            var stream = File.Create(context.FileListPath);
+            using var writer = new StreamWriter(stream) { AutoFlush = true };
+            writer.WriteLine(string.Join(Environment.NewLine, filePathList));
+            stream.Flush();
```

Comment:
FYI I had to do this with the other `StreamWriter` because the `MemoryStream` was used before it was disposed. Did the more obvious thing of giving those two objects separate lifetimes. Cleaned this instance up.